tests/app/main/test_services.py

```diff
@@ -50,129 +51,289 @@ def supplier_remove_service__post_data(request):
+
+            assert res.status_code == 404
+
+    @pytest.mark.parametrize("suffix", ("", "/blah/123",))
+    def test_fails_if_unknown_service(
+            self,
+            data_api_client,
+            service_status,
+            framework_framework,
+            suffix,
+    ):
+        with self.app.test_client():
+            self.login()
+
+            data_api_client.get_service.side_effect = partial(
+                self._mock_get_service_side_effect,
+                service_status,
+                framework_framework,
+                True,
+            )
+            res = self.client.get('/suppliers/services/31415' + suffix)
+
+            assert res.status_code == 404
+
+    @pytest.mark.parametrize("service_belongs_to_user", (False, True,))
+    def test_trailing_slash(
+            self,
+            data_api_client,
+            service_status,
+            framework_framework,
+            service_belongs_to_user,
+    ):
+        with self.app.test_client():
+            self.login()
+
+            data_api_client.get_service.side_effect = partial(
+                self._mock_get_service_side_effect,
+                service_status,
+                framework_framework,
+                True,
+            )
+            res = self.client.get('/suppliers/services/567/')
+
+            assert res.status_code == 301
+            assert res.location == "http://localhost/suppliers/services/567"
+
+
 class TestListServices(BaseApplicationTest):
+    def _mock_get_framework_side_effect(self, framework_slug):
+        try:
+            return {
+                "g-cloud-909": self.framework(status="live", name="G-Cloud 909", slug="g-cloud-909"),
+                "g-cloud-890": self.framework(status="open", name="G-Cloud 890", slug="g-cloud-890"),
+                "g-cloud-870": self.framework(status="expired", name="G-Cloud 870", slug="g-cloud-870"),
+            }[framework_slug]
+        except KeyError:
+            raise HTTPError(mock.Mock(status_code=404))
+
     @mock.patch('app.main.views.services.data_api_client')
     def test_shows_no_services_message(self, data_api_client):
         with self.app.test_client():
             self.login()
 
+            data_api_client.get_framework.side_effect = self._mock_get_framework_side_effect
             data_api_client.find_services.return_value = {
-                "services": []
+                "services": [],
             }
 
-            res = self.client.get('/suppliers/services')
+            res = self.client.get('/suppliers/frameworks/g-cloud-909/services')
             assert res.status_code == 200
+
+            document = html.fromstring(res.get_data(as_text=True))
+            assert document.xpath(
+                "//*[contains(@class, 'summary-item-no-content')][normalize-space(string())=$t]",
+                t=u"You don’t have any G-Cloud 909 services on the Digital Marketplace",
+            )
+            assert document.xpath(
```

Comment:
Nitpick but it feels more logical to assert your way down the visual page i.e starting with the heading and then getting to the table